sam_view.c

```diff
@@ -746,47 +840,16 @@ static bool bam1_to_fq(const bam1_t *b, kstring_t *linebuf, const bam2fq_state_t
         }
     }
     kputc('\n', linebuf);
-
-    seq = bam_get_seq(b);
-
-    if (b->core.flag & BAM_FREVERSE) { // read is reverse complemented
-        for (i = qlen-1; i > -1; --i) {
-            char c = seq_nt16_str[seq_comp_table[bam_seqi(seq,i)]];
-            kputc(c, linebuf);
-        }
-    } else {
-        for (i = 0; i < qlen; ++i) {
-            char c = seq_nt16_str[bam_seqi(seq,i)];
-            kputc(c, linebuf);
-        }
-    }
+    kputs(seq, linebuf);
     kputc('\n', linebuf);
 
     if (state->filetype == FASTQ) {
         // Write quality
         kputs("+\n", linebuf);
-        if (has_qual) {
-            if (state->use_oq && oq) {
-                if (b->core.flag & BAM_FREVERSE) { // read is reverse complemented
-                    for (i = qlen-1; i > -1; --i) {
-                        kputc(oq[i], linebuf);
-                    }
-                } else {
-                    kputs((char*)oq, linebuf);
-                }
-            } else {
-                if (b->core.flag & BAM_FREVERSE) { // read is reverse complemented
-                    for (i = qlen-1; i > -1; --i) {
-                        kputc(33 + qual[i], linebuf);
-                    }
-                } else {
-                    for (i = 0; i < qlen; ++i) {
-                        kputc(33 + qual[i], linebuf);
-                    }
-                }
-            }
+        if (qual && *qual) {
+            kputs(qual, linebuf);
         } else {
-            for (i = 0; i < qlen; ++i) {
+            for (i = 0; i < strlen(seq); ++i) {
```

Comment:
Although modern compilers now optimise this through C extensions like gcc ```__attribute_pure__``` I am unsure of how widespread this and equivalent optimisations are.   Most likely it's standard now, but it still makes me nervous.  Without it, this loop turns from O(N) into O(N^2) complexity.  I would suggest doing strlen once and referring to that within the loop.

Edit: even with gcc 6 and clang 3.8 this optimisation is only made with -O so in a traditional debugging environment the code would become O(N^2).